sdk/iot/core/src/az_iot_telemetry.c

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include <stdint.h>
+
+#include "az_iot_hub_client.h"
+#include <az_precondition.h>
+#include <az_result.h>
+#include <az_span.h>
+
+#include <_az_cfg_prefix.h>
+
+static const uint8_t telemetry_prop_delim = '?';
+static const uint8_t telemetry_prop_separator = '&';
+static const uint8_t telemetry_null_terminator = '\0';
+static const az_span telemetry_topic_prefix = AZ_SPAN_LITERAL_FROM_STR("devices/");
+static const az_span telemetry_topic_modules_mid = AZ_SPAN_LITERAL_FROM_STR("/modules/");
+static const az_span telemetry_topic_suffix = AZ_SPAN_LITERAL_FROM_STR("/messages/events/");
+
+AZ_NODISCARD az_result az_iot_hub_client_telemetry_publish_topic_get(
+    az_iot_hub_client const* client,
+    az_iot_hub_client_properties const* properties,
+    az_span mqtt_topic,
+    az_span* out_mqtt_topic)
+{
+  AZ_PRECONDITION_NOT_NULL(client);
+  AZ_PRECONDITION_VALID_SPAN(mqtt_topic, 0);
+  AZ_PRECONDITION_NOT_NULL(out_mqtt_topic);
+
+  // Required topic parts
+  int32_t required_size = az_span_length(telemetry_topic_prefix)
+      + az_span_length(client->_internal.device_id) + az_span_length(telemetry_topic_suffix)
+      + sizeof(telemetry_null_terminator);
+
+  // Optional parts
+  if (az_span_ptr(client->_internal.options.module_id) != NULL)
+  {
+    required_size += az_span_length(telemetry_topic_modules_mid);
+    required_size += az_span_length(client->_internal.options.module_id);
+  }
+
+  if (properties != NULL)
+  {
+    required_size
+        += az_span_length(properties->_internal.properties) + sizeof(telemetry_prop_delim);
+  }
+
+  if (az_span_ptr(client->_internal.options.user_agent) != NULL)
```

Comment:
is it an option to calculate this required_size when client is init?  and then update it every time it gets updated?
so then we don't have to calculate it every time this publish method is called?
Basically adding another field to the client to know about this at all time

Reply:
this will often be used only once by the developer. They will get their topic that they publish to and reuse that every time they send telemetry. So in the common case the required size will only be calculated once.